### PR TITLE
holding for 3704: ENCD-3705 limit collection

### PR DESCRIPTION
**Pull request by forresttanaka**
The key to this ticket is in the FacetChart component which now splits the facet terms into individual and “other” sections, if the number of terms exceeds the MAX_CHART_ITEMS constant. I store the values and labels as properties of the FacetChart component so that they’re accessible from the callbacks that actually draw the charts. Note that I _could_ store them as React component state too, but changing React state triggers a redraw, which would be unnecessary because these component properties are needed to draw the charts in the first place.

Changing MAX_CHART_ITEMS alone and rebuilding changes the number of items displayed before “other.”

Added two nested functions to the componentDidUpdate() method to generate query strings because both the legend-rendering callback and the doughnut-rendering click callback need to generate the same query strings under the same circumstances.

### CODE REVIEW

File: src/encoded/static/components/collection.js [hunk 1]
```diff
@@ -76,7 +87,17 @@ class FacetChart extends React.Component {
                             const text = [];
                             text.push('<ul>');
                             for (let i = 0; i < chartData.length; i += 1) {
-                                const searchUri = `${baseSearchUri}&${facet.field}=${encodeURIComponent(chartLabels[i]).replace(/%20/g, '+')}`;
+                                let queryString = '';
+                                if (this.otherTerms.length && this.otherTermsIndex === i) {
+                                    // Handle a click in the "other" term. We need to generate a
+                                    // query string that combines all the "other" terms.
+                                    queryString = generateMultipleQuery(facet.field, this.otherTerms);
```
Comment (caseylitton) [inline src/encoded/static/components/collection.js]: Is this dependent on the loop?  It looks like 
```
queryString = generateMultipleQuery(facet.field, this.otherTerms);
```
could be called once when the chart is created since otherTerms is set in render.

File: src/encoded/static/components/collection.js [hunk 2]
```diff
@@ -88,12 +109,23 @@ class FacetChart extends React.Component {
...
                             // React to clicks on pie sections
                             const activePoints = this.chartInstance.getElementAtEvent(e);
                             if (activePoints[0]) {
                                 const clickedElementIndex = activePoints[0]._index;
                                 const chartLabels = this.chartInstance.data.labels;
-                                const searchUri = `${baseSearchUri}&${facet.field}=${encodeURIComponent(chartLabels[clickedElementIndex]).replace(/%20/g, '+')}`;
+                                if (this.otherTerms.length && this.otherTermsIndex === clickedElementIndex) {
+                                    // The click was in the "other" term. We need to generate a
+                                    // query string that combines all the "other" terms.
+                                    queryString = generateMultipleQuery(facet.field, this.otherTerms);
```
Comment (caseylitton) [inline src/encoded/static/components/collection.js]: Same as above, can this be called once when the chart is created?

File: src/encoded/static/components/collection.js [hunk 3]
```diff
@@ -109,22 +141,58 @@ class FacetChart extends React.Component {
...
+
+            // Keep track of the "other" terms and its location (stored in this.otherTermsIndex) in
+            // the values/labels arrays so that the callbacks to handle clicks in the legend and
+            // the doughnut slices can handle the "other" term.
+            this.otherTerms = otherTerms;
+            this.otherTermsIndex = this.labels.length - 1;
+        } else {
+            // Remember we don't have enough terms to need an "other" term.
+            this.otherTerms = [];
+            this.otherTermsIndex = 0;
```
Comment (caseylitton) [inline src/encoded/static/components/collection.js]: Same as above, I know what's going on but it's a confusing code block since 'otherTermsIndex' is not an index for 'otherTerms'.  I'm not telling you to change your variable name but when I first saw this I thought, 'Why are we setting the index variable to 0 when the array is empty?'

File: src/encoded/static/components/collection.js [hunk 4]
```diff
@@ -109,22 +141,58 @@ class FacetChart extends React.Component {
...
+        }
+
         // Check whether we have usable values in one array or the other we just collected (we'll
-        // just use `this;values` here) to see if we need to render a chart or not.
+        // just use `this.values` here) to see if we need to render a chart or not.
```
Comment (caseylitton) [inline src/encoded/static/components/collection.js]: Unnecessary changes!

File: src/encoded/static/components/collection.js [hunk 5]
```diff
@@ -109,22 +141,58 @@ class FacetChart extends React.Component {
...
+        // as another data item to the array of values.
+        if (otherTerms.length) {
+            this.values.push(otherTerms.reduce((sum, term) => sum + term.doc_count, 0));
+            this.labels.push('other');
+
+            // Keep track of the "other" terms and its location (stored in this.otherTermsIndex) in
+            // the values/labels arrays so that the callbacks to handle clicks in the legend and
+            // the doughnut slices can handle the "other" term.
+            this.otherTerms = otherTerms;
+            this.otherTermsIndex = this.labels.length - 1;
```
[the remark this answers is not in the document]
Comment (hitz) [reply src/encoded/static/components/collection.js]: maybe indexOfOther?  Just for future reference.

**Comment by forresttanaka** [comment]
Discussed with Ben, and I’m certain your optimizations are correct (I haven’t looked in detail at it, but it feels like you’re probably correct), and Ben suggests I just look at that again for ENCD-3704 so that we avoid another QA cycle for this one. Ben added a comment to ENCD-3704 to remind me about it. This ticket exists to prepare for ENCD-3704.

On the one-character comment change, that’s based on our summer interns being led astray by typos in comments because none of the interns we’ve had were strong at Javascript, so they thought anything in a comment was gospel and would spend time researching what something like `this;values` meant. I’d really like to correct these when found around code I’m modifying. I don’t know if we’ll get summer interns again, but we’ve had them two years running.

**Comment by caseylitton** [comment]
Okay, let me know which needs review and when to merge one or the other.   

I like marking the unnecessary changes, not to revert them, just to make sure they we're all on the same page.  I didn't know we had summer interns...reviewing their code should be fun.

**Comment by forresttanaka** [comment]
We’re looking at merging this one, then once ENCD-3704 gets scheduled and I start work on it, to revisit your optimizations since I’ll be working in roughly the same code (certainly the same file). Really, this code isn’t very important right now in that very few people would care if it exists or not, much less optimized or not. But it’ll become more important once ENCD-3704 becomes real.

File: src/encoded/static/components/collection.js [hunk 6]
```diff
@@ -109,22 +141,58 @@ class FacetChart extends React.Component {
...
+        // If we have more facet terms than we allow to be displayed (determined through
+        // MAX_CHART_ITEMS), then just get MAX_CHART_ITEMS to display as usual, and collect the
+        // rest to display as "other."
+        if (facet.terms.length > MAX_CHART_ITEMS) {
+            // We have more facet terms than we display individually. First get the terms to
+            // display as individual items.
+            terms = facet.terms.slice(0, MAX_CHART_ITEMS);
+
+            // Now collect the terms to display as "other."
+            otherTerms = facet.terms.slice(MAX_CHART_ITEMS);
```
Comment (davisagli) [inline src/encoded/static/components/collection.js]: If there are 10 items we'll end up with an "other" wedge that only represents a single item, which seems non-ideal. So maybe slice at MAX_CHART_ITEMS - 1 (which basically reinterprets MAX_CHART_ITEMS as the max *including* the "other" item).

File: src/encoded/static/components/collection.js [hunk 7]
```diff
@@ -109,22 +141,58 @@ class FacetChart extends React.Component {
...
 
         // Extract the arrays of labels from the facet keys, and the arrays of corresponding counts
         // from the facet doc_counts. Only use non-zero facet terms in the charts. IF we have no
         // usable data, both these arrays have no entries. componentDidMount assumes these arrays
         // have been populated.
         this.values = [];
         this.labels = [];
-        facet.terms.forEach((term) => {
+        terms.forEach((term) => {
             if (term.doc_count) {
```
Comment (davisagli) [inline src/encoded/static/components/collection.js]: Hmm, so if there are terms with no doc_count we can end up with a chart that has fewer than MAX_CHART_ITEMS wedges but still has an "other"...

**Comment by forresttanaka** [comment]
Thanks @davisagli. That’s good stuff. @caseylitton at this stage, as I don’t think it’s worth spending time on this ticket as it’s just some prep work for ENCD-3704, I recommend closing this, setting ENCD-3705 to in progress, and schedule for the same release as ENCD-3704 if @hitz agrees.
